boards/tests/test_views.py

```diff
@@ -58,3 +60,48 @@ def test_board_detail_view_produces_status_code_200(self):
         request = self.factory.get(reverse("detail", args=(self.board_obj.slug,)))
         response = board_detail(request, self.board_obj.slug)
         self.assertEquals(200, response.status_code)
+
+
+class AddNewTopicTest(TestCase):
+    def setUp(self):
+        self.board_obj = Board.objects.create(
+            name="board", slug="boarddslug", description="1"
+        )
+        self.user = User.objects.create_user(
+            username="sourabh", email="so@dj.com", password="Somepassword"
+        )
+        self.factory = RequestFactory()
+
+    def test_csrf_token_is_present_in_get_request_to_CreateTopicView(self):
+        kwargs = {"slug": self.board_obj.slug}
+        request = self.factory.get(reverse("new_topic", args=(self.board_obj.slug,)))
+        request.user = self.user
+        response = CreateTopicView.as_view()(request, **kwargs)
+        self.assertContains(response, "csrfmiddlewaretoken")
+
+    def test_CreateTopicView_redirects_on_valid_post_data(self):
+        kwargs = {"slug": self.board_obj.slug}
+        url = reverse("new_topic", args=(self.board_obj.slug,))
+        data = {"subject": "Test title", "message": "Lorem ipsum dolor sit amet"}
+        request = self.factory.post(url, data)
+        request.user = self.user
+        response = CreateTopicView.as_view()(request, **kwargs)
+        self.assertEquals(302, response.status_code)
+
+    def test_should_create_topic_with_valid_post_data(self):
+        kwargs = {"slug": self.board_obj.slug}
+        url = reverse("new_topic", args=(self.board_obj.slug,))
+        data = {"subject": "Test title", "message": "Lorem ipsum dolor sit amet"}
+        request = self.factory.post(url, data)
+        request.user = self.user
+        response = CreateTopicView.as_view()(request, **kwargs)
+        self.assertTrue(Topic.objects.exists())
+        self.assertTrue(Post.objects.exists())
+
+    def test_should_render_error_on_no_form_data(self):
+        url = reverse("new_topic", kwargs={"slug": self.board_obj.slug})
+        data = {}
+        self.client.force_login(self.user)
+        response = self.client.post(url, data)
+        form = response.context.get("form")
+        self.assertTrue(form.errors)
```

Comment:
To test context data, we have to go through middlewares,so I used client. This is according to this [stackoverflow answer](https://stackoverflow.com/a/11553035/14354318), when I tried to use it using request factory I got error "Template response has no attribute context"

Reply:
Then this shouldn't be tested at view level. Add test case for the form.